shared/src/commonMain/kotlin/org/hyperskill/app/first_problem_onboarding/presentation/FirstProblemOnboardingReducer.kt

```diff
@@ -0,0 +1,138 @@
+                    null
+                }
+        }
+
+    private fun handleNextLearningActivityResult(
+        state: State,
+        message: FirstProblemOnboardingFeature.FetchNextLearningActivityResult
+    ): FirstProblemOnboardingReducerResult =
+        when (message) {
+            FirstProblemOnboardingFeature.FetchNextLearningActivityResult.Error ->
+                state
+                    .copy(isLearningActivityLoading = false)
+                    .updateNextLearningActivityState(NextLearningActivityState.Error) to
+                    buildSet {
+                        if (state.isLearningActivityLoading) {
+                            add(Action.ViewAction.ShowNetworkError)
+                        }
+                    }
+            is FirstProblemOnboardingFeature.FetchNextLearningActivityResult.Success ->
+                if (state.isLearningActivityLoading) {
+                    state.copy(isLearningActivityLoading = false) to setOf(
+                        getNavigateActionByLearningActivity(message.nextLearningActivity)
+                    )
+                } else {
+                    state.updateNextLearningActivityState(
+                        NextLearningActivityState.Content(message.nextLearningActivity)
+                    ) to emptySet()
+                }
+        }
+
+    private fun handleLearningActionButtonClicked(state: State): FirstProblemOnboardingReducerResult? =
+        if (state.profileState is ProfileState.Content) {
+            val actions = when (state.nextLearningActivityState) {
+                NextLearningActivityState.Idle, NextLearningActivityState.Loading ->
+                    emptySet()
+                NextLearningActivityState.Error ->
+                    setOf(InternalAction.FetchNextLearningActivity)
+                is NextLearningActivityState.Content ->
+                    setOf(
+                        getNavigateActionByLearningActivity(state.nextLearningActivityState.nextLearningActivity),
+                    )
+            }
+
+            state.copy(
+                isLearningActivityLoading = state.nextLearningActivityState is NextLearningActivityState.Error ||
+                    state.nextLearningActivityState is NextLearningActivityState.Loading
+            ) to actions + setOf(
+                InternalAction.LogAnalyticsEvent(FirstProblemOnboardingClickedLearningActionHyperskillAnalyticsEvent)
+            )
+        } else {
+            null
+        }
+
+    private fun handleViewedEvent(state: State): FirstProblemOnboardingReducerResult =
+        state to setOf(
+            InternalAction.SetFirstProblemOnboardingShownFlag,
+            InternalAction.LogAnalyticsEvent(FirstProblemOnboardingViewedHyperskillAnalyticsEvent)
+        )
+
+    private fun getNavigateActionByLearningActivity(learningActivity: LearningActivity?) =
+        Action.ViewAction.CompleteFirstProblemOnboarding(
+            learningActivity?.targetId?.let { stepId ->
```

Comment:
to prevent possible bugs we can also check target type `learningActivity.targetType == LearningActivityTargetType.STEP`